src/testSuiteManager.ts

```diff
@@ -0,0 +1,211 @@
+    return this.getTestItemInternal(log, testId, true, onItemCreated)!
+  }
+
+  /**
+   * Gets a TestItem from the list of tests
+   * @param testId ID of the TestItem to get
+   * @returns TestItem if found, else undefined
+   */
+  public getTestItem(testId: string): vscode.TestItem | undefined {
+    let log = this.log.getChildLogger({label: 'getTestItem'})
+    return this.getTestItemInternal(log, testId, false)
+  }
+
+  /**
+   * Takes a test ID from the test runner output and normalises it to a consistent format
+   *
+   * - Removes leading './' if present
+   * - Removes leading test dir if present
+   */
+  public normaliseTestId(testId: string): string {
+    let log = this.log.getChildLogger({label: `${this.normaliseTestId.name}`})
+    if (testId.startsWith(`.${path.sep}`)) {
+      testId = testId.substring(2)
+    }
+    if (testId.startsWith(this.config.getRelativeTestDirectory())) {
+      testId = testId.replace(this.config.getRelativeTestDirectory(), '')
+    }
+    if (testId.startsWith(path.sep)) {
+      testId = testId.substring(1)
+    }
+    log.debug('Normalised ID: %s', testId)
+    return testId
+  }
+
+  private testIdToUri(testId: string): vscode.Uri {
+    return vscode.Uri.file(path.resolve(this.config.getAbsoluteTestDirectory(), testId.replace(/\[.*\]/, '')))
+  }
+
+  /**
+   * Creates a TestItem and adds it to a TestItemCollection
+   * @param collection
+   * @param testId
+   * @param label
+   * @param uri
+   * @param canResolveChildren
+   * @returns
+   */
+  private createTestItem(
+    testId: string,
+    label: string,
+    parent?: vscode.TestItem,
+    onItemCreated: TestItemCallback = (_) => {},
+    canResolveChildren: boolean = true,
+  ): vscode.TestItem {
+    let log = this.log.getChildLogger({ label: `${this.createTestItem.name}` })
+    let uri = this.testIdToUri(testId)
+    log.debug('Creating test item', {label: label, parentId: parent?.id, canResolveChildren: canResolveChildren, uri: uri})
+    let item = this.controller.createTestItem(testId, label, uri)
+    item.canResolveChildren = canResolveChildren;
+    (parent?.children || this.controller.items).add(item);
+    log.debug('Added test: %s', item.id)
+    onItemCreated(item)
+    return item
+  }
+
+  /**
+   * Splits a test ID into an array of all parent IDs to reach the given ID from the test tree root
+   * @param testId test ID to split
+   * @returns array of test IDs
+   */
+  private getParentIdsFromId(testId: string): string[] {
+    let log = this.log.getChildLogger({label: `${this.getParentIdsFromId.name}`})
+    testId = this.normaliseTestId(testId)
+
+    // Split path segments
+    let idSegments = testId.split(path.sep)
+    log.debug('id segments', idSegments)
+    if (idSegments[0] === "") {
+      idSegments.splice(0, 1)
+    }
+    for (let i = 1; i < idSegments.length - 1; i++) {
+      let precedingSegments = idSegments.slice(0, i + 1)
+      idSegments[i] = path.join(...precedingSegments)
+    }
```

Comment:
There's a bug here because `idSegments` is being modified as we go, so we end up duplicating the segments over and over.

```suggestion
    // Split path segments
    let originalSegments = testId.split(path.sep)
    log.debug('originalSegments', originalSegments)
    if (originalSegments[0] === "") {
      originalSegments.splice(0, 1)
    }

    let idSegments = [...originalSegments]
    for (let i = 1; i < idSegments.length - 1; i++) {
      let precedingSegments = originalSegments.slice(0, i + 1)
      idSegments[i] = path.join(...precedingSegments)
    }
```

Reply:
Nice, thanks very much for the fix! :D